ffi/src/scan.rs

```diff
@@ -0,0 +1,392 @@
+    key: KernelStringSlice,
+    allocate_fn: AllocateStringFn,
+) -> *mut c_void {
+    asbox!(raw_map as boxed_map => {
+        let string_key = String::try_from_slice(key);
+        match boxed_map.values.get(&string_key) {
+            Some(v) => {
+                let slice: KernelStringSlice = v.as_str().into();
+                allocate_fn(slice)
+            }
+            None => std::ptr::null_mut(),
+        }
+    })
+}
+
+/// Get a selection vector out of a [`CDvInfo`] struct
+///
+/// # Safety
+/// Engine is responsible for providing valid pointers for each argument
+#[no_mangle]
+pub unsafe extern "C" fn selection_vector_from_dv(
+    raw_info: *mut CDvInfo,
+    extern_engine_interface: *const ExternEngineInterfaceHandle,
+    state: *mut GlobalScanState,
+) -> *mut KernelBoolSlice {
+    asbox!(raw_info as boxed_info => {
+        asbox!(state as boxed_state => {
+            let extern_engine_interface = unsafe { ArcHandle::clone_as_arc(extern_engine_interface) };
+            let root_url = Url::parse(&boxed_state.kernel_state.table_root).unwrap();
+            let vopt = boxed_info
+                .dv_info
+                .get_selection_vector(extern_engine_interface.table_client().as_ref(), &root_url)
+                .unwrap();
+            match vopt {
+                Some(v) => Box::into_raw(Box::new(v.into())),
+                None => std::ptr::null_mut(),
+            }
+        })
+    })
+}
+
+// Wrapper function that gets called by the kernel, transforms the arguments to make the ffi-able,
+// and then calls the ffi specified callback
+fn rust_callback(
+    context: &mut ContextWrapper,
+    path: &str,
+    size: i64,
+    dv_info: DvInfo,
+    partition_values: HashMap<String, String>,
+) {
+    let path_slice: KernelStringSlice = path.into();
+    let dv_handle = BoxHandle::into_handle(CDvInfo { dv_info });
+    let partition_map_handle = BoxHandle::into_handle(CStringMap {
+        values: partition_values,
+    });
+    (context.callback)(
+        context.engine_context,
+        path_slice,
+        size,
+        dv_handle,
+        partition_map_handle,
+    );
+    unsafe {
+        BoxHandle::drop_handle(dv_handle);
+    }
+}
+
+// Wrap up stuff from C so we can pass it through to our callback
+struct ContextWrapper {
+    engine_context: *mut c_void,
+    callback: CScanCallback,
+}
+
+/// Shim for ffi to call visit_scan_data. This will generally be called when iterating through scan
+/// data which provides the data handle and selection vector as each element in the iterator.
+///
+/// # Safety
+/// engine is responsbile for passing a valid [`EngineDataHandle`] and selection vector.
+#[no_mangle]
+pub unsafe extern "C" fn visit_scan_data(
+    data: *mut EngineDataHandle,
+    vector: KernelBoolSlice,
+    engine_context: *mut c_void,
+    callback: CScanCallback,
+) {
+    let selection_vec = vector.make_vec();
+    let data: &dyn EngineData = unsafe { (*data).data.as_ref() };
```

Comment:
to minimize scope of `unsafe`, suggest:
```suggestion
    let data = unsafe { &*data };
    let data = data.data.as_ref();
```